File: library/src/main/java/com/imangazaliev/circlemenu/MenuController.java
package com.imangazaliev.circlemenu;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.widget.Toast;

import java.util.List;

class MenuController {

    private static final int TOGGLE_ANIMATION_DURATION = 200;

    private List<CircleMenuButton> buttons;
    private float menuCenterX, menuCenterY;
    private float startAngle;
    private float angleRange;
    private int distance;
    private boolean isOpened;

    private final ItemSelectionAnimator itemSelectionAnimator;
    private final MenuControllerListener listener;


    MenuController(Context context,
                   List<CircleMenuButton> menuButtons,
                   final MenuControllerListener listener,
                   float menuCenterX,
                   float menuCenterY,
                   final float startAngle,
                   final float angleRange,
                   int distance,
                   boolean openOnStart,
                   final boolean hintsEnabled) {
        this.buttons = menuButtons;
        this.menuCenterX = menuCenterX;
        this.menuCenterY = menuCenterY;
        this.startAngle = startAngle;
        this.angleRange = angleRange;
        this.distance = distance;
        this.isOpened = openOnStart;
        this.listener = listener;
        this.itemSelectionAnimator = new ItemSelectionAnimator(context, this, listener, menuCenterX, menuCenterY, distance);

        View.OnClickListener onButtonItemClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                CircleMenuButton menuButton = (CircleMenuButton) v;
                if (menuButton.isShowClickAnim()) {
                    float buttonAngle = (angleRange / buttons.size()) * buttons.indexOf(menuButton) + startAngle;
                    itemSelectionAnimator.startSelectAnimation(menuButton, buttonAngle);
                }
                listener.onClick(menuButton);
            }
        };
        View.OnLongClickListener onButtonItemLongClickListener = new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                CircleMenuButton menuButton = (CircleMenuButton) v;
                if (hintsEnabled) {
                    Toast.makeText(menuButton.getContext(), menuButton.getHintText(), Toast.LENGTH_SHORT).show();
                }
                return hintsEnabled;
            }
        };

        for (CircleMenuButton menuButton : buttons) {
            menuButton.setOnClickListener(onButtonItemClickListener);
            menuButton.setOnLongClickListener(onButtonItemLongClickListener);
        }

        layoutButtons(openOnStart ? distance : 0);
    }

    void onDraw(Canvas canvas) {
        itemSelectionAnimator.onDraw(canvas);
    }

    void toggle() {
        if (isOpened()) {
            close(true);
        } else {
            open(true);
        }
    }

    void open(boolean animate) {
        if (isOpened) {
            return;
        }

        enableButtons(false);
        layoutButtons(0);
        showButtons(true);
        listener.onOpenAnimationStart();

        ValueAnimator buttonAnimator = ValueAnimator.ofFloat(0f, distance);
        buttonAnimator.setDuration(animate ? TOGGLE_ANIMATION_DURATION : 0);
        buttonAnimator.setInterpolator(new DecelerateInterpolator());
        buttonAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                layoutButtons((float) animation.getAnimatedValue());
            }
        });
        buttonAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                isOpened = true;
                enableButtons(true);
                listener.onOpenAnimationEnd();
            }
        });
        buttonAnimator.start();
    }

    void close(boolean animate) {
        if (!isOpened) {
            return;
        }

        enableButtons(false);
        layoutButtons(distance);
        listener.onCloseAnimationStart();

        ValueAnimator buttonAnimator = ValueAnimator.ofFloat(distance, 0f);
        buttonAnimator.setDuration(animate ? TOGGLE_ANIMATION_DURATION : 0);
        buttonAnimator.setInterpolator(new DecelerateInterpolator());
        buttonAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                layoutButtons((float) animation.getAnimatedValue());
            }
        });
        buttonAnimator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                isOpened = false;
                showButtons(false);
                listener.onCloseAnimationEnd();
            }
        });
        buttonAnimator.start();
    }

    private void layoutButtons(float distance) {
        int buttonsCount = buttons.size();
        float angleStep = angleRange / buttonsCount;
        float lastAngle = startAngle;

        for (int i = 0; i < buttonsCount; i++) {
            CircleMenuButton button = buttons.get(i);

            float x = Math.round((float) (menuCenterX + distance * Math.cos(Math.toRadians(lastAngle))));
            float y = Math.round((float) (menuCenterY + distance * Math.sin(Math.toRadians(lastAngle))));

            button.setX(x);
            button.setY(y);

            if (lastAngle > startAngle + angleRange) {
                lastAngle -= angleRange;
            }

            lastAngle += angleStep;
        }

    }

    void setOpened(boolean isOpened) {
        this.isOpened = isOpened;
    }

    boolean isOpened() {
        return isOpened;
    }

    void enableButtons(boolean enabled) {
        for (CircleMenuButton button : buttons) {
            button.setEnabled(enabled);
        }
    }

    void showButtons(boolean visible) {
        for (CircleMenuButton button : buttons) {
            button.setVisibility(visible ? View.VISIBLE : View.INVISIBLE);
        }
    }

}

File: library/src/main/java/com/imangazaliev/circlemenu/MenuController.kt
package com.imangazaliev.circlemenu

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.view.View
import android.view.View.OnLongClickListener
import android.view.animation.DecelerateInterpolator
import kotlin.math.cos
import kotlin.math.roundToInt
import kotlin.math.sin

internal class MenuController(
        context: Context?,
        private val buttons: List<CircleMenuButton>,
        private val listener: Listener,
        private val startAngle: Float,
        private val maxAngle: Float,
        private val distance: Int,
        private val showSelectAnimation: Boolean,
        var isOpened: Boolean
) {

    companion object {
        private const val TOGGLE_ANIMATION_DURATION = 200
    }

    private var menuCenterX: Float = 0f
    private var menuCenterY: Float = 0f
    private val itemSelectionAnimator: ItemSelectionAnimator = ItemSelectionAnimator(
            context = context!!,
            menuController = this,
            controllerListener = listener,
            circleRadius = distance
    )

    init {
        val onButtonItemClickListener = View.OnClickListener { v ->
            val menuButton = v as CircleMenuButton
            if (showSelectAnimation) {
                val buttonAngle = maxAngle / buttons.size * buttons.indexOf(menuButton) + startAngle
                val buttonIndex = buttons.indexOf(menuButton)
                itemSelectionAnimator.startSelectAnimation(menuButton, buttonIndex, buttonAngle)
            } else {
                close(true)
            }
            listener.onButtonClick(menuButton, buttons.indexOf(menuButton))
        }
        val onButtonItemLongClickListener = OnLongClickListener { v ->
            val menuButton = v as CircleMenuButton
            listener.onButtonLongClick(menuButton, buttons.indexOf(menuButton))
            true
        }
        for (menuButton in buttons) {
            menuButton.setOnClickListener(onButtonItemClickListener)
            menuButton.setOnLongClickListener(onButtonItemLongClickListener)
        }
        showButtons(isOpened)
        layoutButtons(if (isOpened) distance.toFloat() else 0.toFloat())
    }

    fun onDraw(canvas: Canvas) {
        itemSelectionAnimator.onDraw(canvas)
    }

    fun toggle() {
        if (isOpened) {
            close(true)
        } else {
            open(true)
        }
    }

    fun open(animate: Boolean) {
        if (isOpened) {
            return
        }
        enableButtons(false)
        layoutButtons(0f)
        showButtons(true)
        listener.onOpenAnimationStart()
        val buttonAnimator = ValueAnimator.ofFloat(0f, distance.toFloat())
        buttonAnimator.duration = if (animate) TOGGLE_ANIMATION_DURATION.toLong() else 0.toLong()
        buttonAnimator.interpolator = DecelerateInterpolator()
        buttonAnimator.addUpdateListener { animation -> layoutButtons(animation.animatedValue as Float) }
        buttonAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                isOpened = true
                enableButtons(true)
                listener.onOpenAnimationEnd()
            }
        })
        buttonAnimator.start()
    }

    fun close(animate: Boolean) {
        if (!isOpened) {
            return
        }
        enableButtons(false)
        layoutButtons(distance.toFloat())
        listener.onCloseAnimationStart()
        val buttonAnimator = ValueAnimator.ofFloat(distance.toFloat(), 0f)
        buttonAnimator.duration = if (animate) TOGGLE_ANIMATION_DURATION.toLong() else 0.toLong()
        buttonAnimator.interpolator = DecelerateInterpolator()
        buttonAnimator.addUpdateListener { animation -> layoutButtons(animation.animatedValue as Float) }
        buttonAnimator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                isOpened = false
                showButtons(false)
                listener.onCloseAnimationEnd()
            }
        })
        buttonAnimator.start()
    }

    private fun layoutButtons(distance: Float) {
        val buttonsCount = buttons.size
        val angleStep = maxAngle / buttonsCount
        var lastAngle = startAngle
        for (i in 0 until buttonsCount) {
            val button = buttons[i]
            val x = (menuCenterX + distance * cos(Math.toRadians(lastAngle.toDouble()))).toFloat().roundToInt().toFloat()
            val y = (menuCenterY + distance * sin(Math.toRadians(lastAngle.toDouble()))).toFloat().roundToInt().toFloat()
            button.x = x
            button.y = y
            if (lastAngle > maxAngle) {
                lastAngle -= maxAngle
            }
            lastAngle += angleStep
        }
    }

    fun enableButtons(enabled: Boolean) {
        for (button in buttons) {
            button.isEnabled = enabled
        }
    }

    fun showButtons(visible: Boolean) {
        for (button in buttons) {
            button.visibility = if (visible) View.VISIBLE else View.INVISIBLE
        }
    }

    fun setCenterButtonPosition(centerButtonX: Float, centerButtonY: Float) {
        this.menuCenterX = centerButtonX
        this.menuCenterY = centerButtonY
        itemSelectionAnimator.setCenterButtonPosition(menuCenterX, menuCenterY)
        layoutButtons(if (isOpened) distance.toFloat() else 0.toFloat())
    }

    internal interface Listener {

        fun onButtonClick(menuButton: CircleMenuButton, index: Int)

        fun onButtonLongClick(menuButton: CircleMenuButton, index: Int)

        fun onOpenAnimationStart()

        fun onOpenAnimationEnd()

        fun onCloseAnimationStart()

        fun onCloseAnimationEnd()

        fun onSelectAnimationStart(buttonIndex: Int)

        fun onSelectAnimationEnd(buttonIndex: Int)

        fun redrawView()

    }

}